### ISSUE TITLE
Inconsistent heap dumps

### ISSUE THREAD

**Issue by perlun** [schneems/heapy#11]
Hi,

Again, thanks for a great gem. 👍 I am investigating a probable leak, have been doing so for about a week or more without any real luck yet... but eventually, I will find it. 😄 

One thing that currently puzzles me is that the number of objects, and amount of heap memory being used, varies so greatly depending on where I look.

When I've done a dump using this code:

```ruby
  file = File.open('/tmp/dump.json', 'w')
  GC.start
  ObjectSpace.dump_all(output: file)
  file.close()
```

...and then ran it via `heapy`, I got this output:

```
$ bundle exec bin/heapy read dump.json

Analyzing Heap
==============
Generation:  nil object count: 140137, mem: 0.0 kb
Generation:   53 object count: 31, mem: 12.2 kb
Generation:   54 object count: 5018, mem: 551.0 kb
Generation:   55 object count: 21337, mem: 5383.3 kb
Generation:   56 object count: 18861, mem: 1816.3 kb
Generation:   57 object count: 5087, mem: 503.4 kb
Generation:   58 object count: 18637, mem: 1142.0 kb
Generation:   59 object count: 4873, mem: 4667.3 kb
Generation:   60 object count: 124, mem: 5.5 kb
Generation:   61 object count: 202, mem: 24.1 kb
Generation:   62 object count: 120, mem: 8.9 kb
Generation:   63 object count: 99, mem: 8.4 kb
Generation:   66 object count: 157, mem: 22.4 kb
Generation:   95 object count: 55, mem: 4.0 kb
Generation:   96 object count: 28, mem: 2.2 kb
Generation:   99 object count: 16, mem: 1.3 kb
Generation:  100 object count: 14, mem: 1.0 kb
Generation:  104 object count: 8, mem: 0.6 kb
Generation:  248 object count: 1648, mem: 146.3 kb
Generation:  249 object count: 388, mem: 33.0 kb
Generation:  250 object count: 8, mem: 0.7 kb
Generation:  387 object count: 32, mem: 2.8 kb
Generation:  417 object count: 16, mem: 1.4 kb
Generation:  418 object count: 28, mem: 2.5 kb
Generation:  419 object count: 8, mem: 0.7 kb
Generation:  549 object count: 32, mem: 2.8 kb
Generation:  577 object count: 16, mem: 1.2 kb
Generation:  655 object count: 7, mem: 0.3 kb
Generation:  709 object count: 21, mem: 0.8 kb
Generation:  840 object count: 3, mem: 0.1 kb
Generation:  858 object count: 36, mem: 3.2 kb
Generation:  863 object count: 94, mem: 27.7 kb
Generation:  865 object count: 4, mem: 0.4 kb
Generation:  938 object count: 1, mem: 0.1 kb
Generation:  946 object count: 1, mem: 0.1 kb
Generation: 1018 object count: 16, mem: 1.4 kb
Generation: 1021 object count: 1, mem: 0.0 kb
Generation: 1022 object count: 3940, mem: 283.8 kb
Generation: 1023 object count: 66, mem: 3.8 kb
Generation: 1024 object count: 240, mem: 17.5 kb
Generation: 1025 object count: 937, mem: 68.1 kb
Generation: 1026 object count: 1781, mem: 134.5 kb
Generation: 1027 object count: 24, mem: 1.9 kb
Generation: 1089 object count: 61, mem: 5.5 kb
Generation: 1091 object count: 197, mem: 37.3 kb
Generation: 1092 object count: 4, mem: 0.3 kb
Generation: 1103 object count: 4, mem: 0.3 kb
Generation: 1116 object count: 12, mem: 1.1 kb
Generation: 1117 object count: 24, mem: 2.0 kb
Generation: 1125 object count: 10, mem: 0.4 kb
Generation: 1126 object count: 64, mem: 5.5 kb
Generation: 1127 object count: 991, mem: 2141.3 kb
Generation: 1128 object count: 15, mem: 1.0 kb

Heap total
==============
Generations (active): 53
Count: 225534
Memory: 17083.5 kb
```

**However**, when I looked at the GC stats (`GC.stat`) from roughly the same point in time, it looked like this:

```json
{
  "count": 1127,
  "heap_allocated_pages": 2390,
  "heap_sorted_length": 3695,
  "heap_allocatable_pages": 0,
  "heap_available_slots": 974164,
  "heap_live_slots": 713354,
  "heap_free_slots": 260810,
  "heap_final_slots": 0,
  "heap_marked_slots": 450378,
  "heap_eden_pages": 2390,
  "heap_tomb_pages": 0,
  "total_allocated_pages": 13653,
  "total_freed_pages": 11263,
  "total_allocated_objects": 188993554,
  "total_freed_objects": 188280200,
  "malloc_increase_bytes": 24150864,
  "malloc_increase_bytes_limit": 33554432,
  "minor_gc_count": 1026,
  "major_gc_count": 101,
  "remembered_wb_unprotected_objects": 1406,
  "remembered_wb_unprotected_objects_limit": 2548,
  "old_objects": 430356,
  "old_objects_limit": 445420,
  "oldmalloc_increase_bytes": 92467016,
  "oldmalloc_increase_bytes_limit": 126476361
}
```

`old_object` 430356 is quite a lot more than 225534 being reported in the heap dump via `heapy`. Shouldn't these be roughly the same?

Also, the total memory being used differs: 17083.5 kb above, whereas `ObjectSpace.memsize_of_all` gave a value of 83409077 (is that bytes or kilobytes? Nonetheless, it is differs)

The total RSS size of the Ruby process is at this time 1267 MiB.

(I have done quite a bit of digging trying to find extension-based leaks with `jemalloc` etc., but haven't yet concluded that anything there is proven to be leaking. To the contrary, when I ran some isolated tests trying to reproduce the problems, all the allocations seemed to come from ObjectSpace and other Ruby-related allocations. Of course this _can_ still be an extension leaking but the call graphs hasn't pointed me in that direction; can give more details about this if you like.)

Any suggestions are helpful. I am very much wandering in the dark here, looking for _something_ that looks fishy, but... it's very dark tonight. 🌑  😉

**Comment by schneems** [comment]
heapy only parses what ObjectSpace gives us, make sure you're tracing as soon as possible. The GC.stat will have some extra stuff in there from before ObjectSpace tracing was turned on so i'm not surprised its different.

It might GC some of those objects before you had a chance to start tracing.

You can do a GC.stat before and then GC.stat after and diff the two for a better approximation instead of relying on only one call to GC.stat.

If this is a rails app, 

For finding a leak, i recommend derailed_benchmarks. First step is to reproduce the leak, hit the task in a loop and verify that memory does not level off. In 100% of "memory leaks" that i've investigated none of them were true leaks, i.e. the memory leveled off after awhile and memory growth was large but fairly normal.

I recommend scout for production tracing that will identify endpoints with large amount of memory use. 

I recommend rack mini profiler for development and production. You can add a query param to get it to output the total number of objects that were generated on a specific page.

In general. Reproduce the problem, use debugging tools to isolate the problem, guess and check until the problem goes away.

**Comment by perlun** [comment]
Thanks @schneems. I tried my best at reproducing the potential leak, but it was hard; I was struggling with knowing that I had really reproduced the leak and not just "regular memory usage". The most likely code path for reproducing the path was also quite slow, meaning I couldn't easily run a high number of requests to reliably see if memory usage was growing.

So my only reliable repro was in production. I ended up upgrading to Ruby `trunk`, since I was wanting to compile from source anyway (to enable jemalloc). And the problem seemed to go away! I was randomly googling one morning for "ruby 2.4.1 memory leak" and found https://github.com/rest-client/rest-client/issues/611 which hinted at the underlying MRI issue: [Bug #13772: Memory leak recycling stacks for threads in 2.4.1](https://bugs.ruby-lang.org/issues/13772).

So, I was lucky enough to hit a leak in MRI actually. 😄 Hopefully, the next time, it will be my own code that is at fault instead. 😉 Thanks for a great tool anyhow!
